File: java/src/main/be/steinsomers/bron_kerbosch/study/SampleStatistics.java
package be.steinsomers.bron_kerbosch.study;

final class SampleStatistics {
    private long max;
    private long min;
    private int samples;
    private double sum;
    private double sum_of_squares;

    public void put(final long v) {
        if (samples == 0 || min > v) {
            min = v;
        }
        if (samples == 0 || max < v) {
            max = v;
        }
        samples += 1;
        sum += v;
        sum_of_squares += v * v;
    }

    public long min() {
        return min;
    }

    public long max() {
        return max;
    }

    public double mean() {
        if (samples > 0) {
            return Math.max(min, Math.min(max, sum / samples));
        } else {
            return Double.NaN;
        }
    }

    public double variance() {
        if (samples > 1) {
            final double n = samples;
            return Math.max(sum_of_squares - sum * sum / n, 0) / (n - 1.0);
        } else {
            return Double.NaN;
        }
    }

    public double deviation() {
        return Math.min(max - min, Math.sqrt(variance()));
    }
}

File: kotlin/src/main/kotlin/be/steinsomers/bron_kerbosch/study/SampleStatistics.kt
package be.steinsomers.bron_kerbosch.study

import kotlin.math.sqrt

internal class SampleStatistics {
    private var max = 0L
    private var min = 0L
    private var samples = 0
    private var sum = 0.0
    private var sumOfSquares = 0.0

    fun put(v: Long) {
        if (samples == 0 || min > v) {
            min = v
        }
        if (samples == 0 || max < v) {
            max = v
        }
        samples += 1
        sum += v.toDouble()
        sumOfSquares += (v * v).toDouble()
    }

    fun min(): Long {
        return min
    }

    fun max(): Long {
        return max
    }

    fun mean(): Double {
        return if (samples > 0) {
            kotlin.math.max(min.toDouble(), kotlin.math.min(max.toDouble(), sum / samples))
        } else {
            Double.NaN
        }
    }

    fun variance(): Double {
        if (samples > 1) {
            val n = samples.toDouble()
            return kotlin.math.max(sumOfSquares - sum * sum / n, 0.0) / (n - 1.0)
        } else {
            return Double.NaN
        }
    }

    fun deviation(): Double {
        return kotlin.math.min((max - min).toDouble(), sqrt(variance()))
    }
}
